### PR TITLE
feature: add support for fasthttp

### PR DESCRIPTION
**Pull request by xuyue97**
feature: add support for fasthttp

### CODE REVIEW

File: plugins/fasthttp/client_intercepter.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func (h *ClientInterceptor) AfterInvoke(invocation operator.Invocation, result ...interface{}) error {
+	if invocation.GetContext() == nil {
+		return nil
+	}
+	span := invocation.GetContext().(tracing.Span)
+	if resp, ok := invocation.Args()[1].(*fasthttp.Response); ok && resp != nil {
+		if resp.StatusCode() >= 400 {
+			span.Error(string(resp.Body()))
```
Comment (mrproliu) [inline plugins/fasthttp/client_intercepter.go:55]: Please don't add the response body to the span. The response body could be bytes, streaming, you can just mark the span as error.

File: plugins/fasthttp/go.mod [hunk 2]
```diff
@@ -0,0 +1,15 @@
+module github.com/apache/skywalking-go/plugins/fasthttp
+
+go 1.18
+
+require (
+	github.com/apache/skywalking-go/plugins/core v0.0.0-20231012073724-907c04460023
```
Comment (mrproliu) [inline plugins/fasthttp/go.mod:6]: We already using the go workspace, please delete this package declaration.

File: test/plugins/scenarios/fasthttp/go.mod [hunk 3]
```diff
@@ -0,0 +1,25 @@
+module test/plugins/scenarios/fasthttp
+
+go 1.18
+
+require (
+	github.com/apache/skywalking-go v0.2.0
```
Comment (mrproliu) [inline test/plugins/scenarios/fasthttp/go.mod:6]: Please delete this package declarations.

**Review by mrproliu — COMMENTED** [review]
Please fix the CI and comments.

File: test/plugins/scenarios/fasthttp/plugin.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+entry-service: http://${HTTP_HOST}:${HTTP_PORT}/consumer
+health-checker: http://${HTTP_HOST}:${HTTP_PORT}/health
+start-script: ./bin/startup.sh
+framework: go
+export-port: 8080
+support-version:
+  - go: 1.17
```
Comment (mrproliu) [inline test/plugins/scenarios/fasthttp/plugin.yml:23]: I think you should add more fast http version checkers. Not only the go version.

File: test/plugins/scenarios/fasthttp/config/excepted.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+segmentItems:
+  - serviceName: fasthttp
+    segmentSize: ge 1
+    segments:
+      - segmentId: not null
+        spans:
+          - operationName: GET:http://localhost:8080/provider
```
Comment (mrproliu) [inline test/plugins/scenarios/fasthttp/config/excepted.yml:23]: Could you also add the `Exit` span for checking the span which generated from the plugin?

File: test/plugins/scenarios/fasthttp/main.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+	ctx.SetStatusCode(resp.StatusCode())
+	ctx.Write(resp.Body())
+
+	fasthttp.ReleaseRequest(req)
+	fasthttp.ReleaseResponse(resp)
+}
+
+func main() {
+	fasthttp.ListenAndServe(":8080", func(ctx *fasthttp.RequestCtx) {
```
Comment (mrproliu) [inline test/plugins/scenarios/fasthttp/main.go:55]: Is the `fasthttp` is HTTP server or client framework? In this demo, it appears to be both a server and a client at the same time. However, your plugin code only provides interception for HTTP clients?

Comment (xuyue97) [reply test/plugins/scenarios/fasthttp/main.go:55]: Thank you for your review.
We do not use the server in our project, so we only write the enhancement to the client. 
I'll add it later.

Comment (mrproliu) [reply test/plugins/scenarios/fasthttp/main.go:55]: Got it, will you add it in this PR or a new one?

Comment (xuyue97) [reply test/plugins/scenarios/fasthttp/main.go:55]: In this PR, I will supplement the code in the next few days

File: plugins/fasthttp/client/client_intercepter_test.go [hunk 7]
```diff
@@ -63,9 +63,9 @@ func TestClientInvoke(t *testing.T) {
 	spans := core.GetReportedSpans()
 	assert.NotNil(t, spans, "spans should not be nil")
 	assert.Equal(t, 1, len(spans), "spans length should be 1")
-	assert.Equal(t, "GET:http://localhost/", spans[0].OperationName(), "operation name should be GET:/")
+	assert.Equal(t, "GET:http://localhost/", spans[0].OperationName(), "operation name should be GET:http://localhost/")
 	assert.Equal(t, "Http", spans[0].SpanLayer().String(), "SpanLayer should be Http")
-	assert.Equal(t, int32(5019), spans[0].ComponentID(), "ComponentID should be 5014")
+	assert.Equal(t, int32(5019), spans[0].ComponentID(), "ComponentID should be 5019")
```
Comment (mrproliu) [inline plugins/fasthttp/client/client_intercepter_test.go:68]: Looks like this component ID did [not exist in the backend](https://github.com/apache/skywalking/blob/master/oap-server/server-starter/src/main/resources/component-libraries.yml). will you create a PR on the main repo?

Comment (xuyue97) [reply plugins/fasthttp/client/client_intercepter_test.go:68]: After merging this PR, I will submit the PR for the main repo: skywalking

Comment (wu-sheng) [reply plugins/fasthttp/client/client_intercepter_test.go:68]: ID should be submitted and merged first.

Comment (wu-sheng) [reply plugins/fasthttp/client/client_intercepter_test.go:68]: Otherwise, others may have registered your ID before your submission.

Comment (xuyue97) [reply plugins/fasthttp/client/client_intercepter_test.go:68]: Okay, I have submitted the PR on the main repo

File: docs/en/agent/support-plugins.md [hunk 8]
```diff
@@ -10,6 +10,7 @@ metrics based on the tracing data.
   * `iris`: [Iris](https://github.com/kataras/iris) tested v12.1.0 to 12.2.5.
 * HTTP Client
   * `http`: [Native HTTP](https://pkg.go.dev/net/http) tested go v1.17 to go v1.20.
+  * `fasthttp`: [FastHttp](https://github.com/valyala/fasthttp) tested go v1.17 to go v1.20.
```
[the remark this answers is not in the document]
Comment (mrproliu) [reply docs/en/agent/support-plugins.md:13]: Please also add this plugin to the HTTP server, and add the tested version declaration.

File: test/plugins/scenarios/fasthttp/config/excepted.yml [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            parentSpanId: -1
+            spanId: 0
+            spanLayer: Http
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 5019
+            isError: false
+            spanType: Exit
```
Comment (mrproliu) [inline test/plugins/scenarios/fasthttp/config/excepted.yml:31]: I think It should be an entry span?

File: docs/en/agent/support-plugins.md [hunk 10]
```diff
@@ -8,9 +8,10 @@ metrics based on the tracing data.
   * `go-restfulv3`: [Go-Restful](https://github.com/emicklei/go-restful) tested v3.7.1 to 3.10.2.
   * `mux`: [Mux](https://github.com/gorilla/mux) tested v1.7.0 to v1.8.0.
   * `iris`: [Iris](https://github.com/kataras/iris) tested v12.1.0 to 12.2.5.
+  * `fasthttp`: [FastHttp](https://github.com/valyala/fasthttp) tested every ten versions from v1.0.0 to v1.50.0.
```
Comment (wu-sheng) [inline docs/en/agent/support-plugins.md:11]: What do you mean every ten version? Do you mean 1.0, 1.10, 1.20?

Comment (xuyue97) [reply docs/en/agent/support-plugins.md:11]: Yes, should i change the instructions here

Comment (wu-sheng) [reply docs/en/agent/support-plugins.md:11]: You could simply say 1.0.0 to 1.50.

Comment (xuyue97) [reply docs/en/agent/support-plugins.md]: Got it.
it has been changed.

File: test/plugins/scenarios/fasthttp/config/excepted.yml (hunk 9 above)
Comment (xuyue97) [reply test/plugins/scenarios/fasthttp/config/excepted.yml]: This file has been modified

File: test/plugins/scenarios/fasthttp/plugin.yml [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+      - v1.41.0
+      - v1.42.0
+      - v1.43.0
+      - v1.44.0
+      - v1.45.0
+      - v1.46.0
+      - v1.47.0
+      - v1.48.0
+      - v1.49.0
+      - v1.50.0
```
Comment (wu-sheng) [inline test/plugins/scenarios/fasthttp/plugin.yml:35]: Why don't test 1.10, 1.20, 1.30, 1.40, 1.50? I think that would be more reliable.

File: test/plugins/scenarios/fasthttp/config/excepted.yml [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+segmentItems:
+  - serviceName: fasthttp
+    segmentSize: ge 3
+    segments:
+      - segmentId: not null
+        spans:
+          - operationName: GET:http://service:8080/health
```
Comment (wu-sheng) [inline test/plugins/scenarios/fasthttp/config/excepted.yml:23]: We don't need to check `health`.

Comment (xuyue97) [reply test/plugins/scenarios/fasthttp/config/excepted.yml]: This error occurred again in the kratosv2 test

actual: 
  Segment[20e6703a6e1d11eeb1e80242ac120003.29.39094317935270002] e
  expected:	Span[-1, 0] /helloworld.Greeter/SayHello
  actual:	span[-1, 0] GET:/health
  reason:	[operation name]: expected=>{/helloworld.Greeter/SayHello}, actual=>{GET:/health}

File: test/plugins/scenarios/fasthttp/plugin.yml (hunk 11 above)
Comment (xuyue97) [reply test/plugins/scenarios/fasthttp/plugin.yml]: Okay, I have modified the configuration

File: test/plugins/scenarios/fasthttp/config/excepted.yml (hunk 12 above)
Comment (wu-sheng) [reply test/plugins/scenarios/fasthttp/config/excepted.yml]: `/health` should not exist, that is first. Then, that test may fail due to something inside that plugin or test case not stable.

Comment (xuyue97) [reply test/plugins/scenarios/fasthttp/config/excepted.yml]: All right, I got it

**Review by mrproliu — APPROVED** [review]

**Pull request merged**